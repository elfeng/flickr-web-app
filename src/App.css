.App {
  text-align: center;
  margin-bottom: 30px;
}

.App-header {
  background-color: black;
  color: white;
  padding-top: 20px;
  padding-bottom: 10px;
  margin-bottom: 30px;
  text-align: left;
}

.App-header h1 {
  font-size: 26px;
}

.App-header h2 {
  padding-top: 5px;
  font-size: 19px;
}

.App-header h1, h2 {
  display: inline;
  text-align: left;
}

.App-form {
  padding-top: 10px;
  padding-bottom: 5px;
}

.App-intro {
  text-align: justify;
  padding-left: 20px;
}

.App-photos {
  text-align: center;
  padding: 0px;
}

.App-intro h2 {
  font-size: 22px;
}

p {
  margin-top: 15px;
  margin-bottom: 20px;
}

.form-control, .btn {
  margin-bottom: 10px;
}
