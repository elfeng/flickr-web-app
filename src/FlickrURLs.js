const API_KEY = 'a5e95177da353f58113fd60296e1d250';
const NASA_USER_ID = '24662369@N07';
const PHOTOS_ONLY = 1;
const SAFE_SEARCH_ACTIVATED = 1;
const NB_MAX_SEARCH_RESULTS = 100;

export const NB_MAX_PHOTOS_PER_PAGE = 20;

export function getSearchUrl(search, month, year) {
    return `https://api.flickr.com/services/rest/?method=flickr.photos.search&
    api_key=${API_KEY}&
    user_id=${NASA_USER_ID}&
    content_type=${PHOTOS_ONLY}&
    per_page=${NB_MAX_SEARCH_RESULTS}&
    format=json&
    nojsoncallback=1&
    sort=relevance&
    safe_search=${SAFE_SEARCH_ACTIVATED}&
    text=${search}&
    min_taken_date=${month.getMinDateAsSeconds(year)}&
    max_taken_date=${month.getMaxDateAsSeconds(year)}`;
};

export function getPhotoInfoUrl(photoId) {
    return `https://api.flickr.com/services/rest/?method=flickr.photos.getInfo&format=json&nojsoncallback=1&api_key=${API_KEY}&photo_id=${photoId}`;
};

export function getPhotoSnapshotUrl({farm, server, id, secret}) {
    return `https://farm${farm}.staticflickr.com/${server}/${id}_${secret}.jpg`;
};
