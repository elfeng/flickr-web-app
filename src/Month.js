export class Month {

    /**
     * number : 1-12
     */
    constructor(number, name){
        this.number = number;
        this.name = name;
    }

    getMinDateAsSeconds(year){
        return new Date(year, this.number - 1, 1).getTime() / 1000;
    }

    getMaxDateAsSeconds(year){
         return new Date(year, this.number, 1).getTime() / 1000;
    }

    isNotAfterCurrentMonth(){
        return this.number <= new Date().getMonth() + 1;
    }
};

export const TWELVE_MONTHS = [
    new Month(1, "January"),
    new Month(2, "February"),
    new Month(3, "March"),
    new Month(4, "April"),
    new Month(5, "May"),
    new Month(6, "June"),
    new Month(7, "July"),
    new Month(8, "August"),
    new Month(9, "September"),
    new Month(10, "October"),
    new Month(11, "November"),
    new Month(12, "December")
];
