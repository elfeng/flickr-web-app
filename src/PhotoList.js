import React from 'react';

export default function PhotoList(props) {

    if (props.nbSearchResults === 0) {
        return (<span>No photo found</span>);

    } else {
        return (
            <div>
                { props.photoContainers }
            </div>
        );
    }
}
