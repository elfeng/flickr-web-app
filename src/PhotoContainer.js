import React, { Component } from 'react';
import axios from 'axios';
import Photo from './Photo/Photo';
import FlickrPhotoWrapper from './FlickrPhotoWrapper';
import { getPhotoInfoUrl } from './FlickrURLs';

export default class PhotoContainer extends Component {

    constructor(props) {
        super(props);
        this.state = {
            photoWrapper: null
        };
        this.getPhotoDetailsFromFlickr = this.getPhotoDetailsFromFlickr.bind(this);
    }

    componentDidMount() {
        const self = this;
        this.getPhotoDetailsFromFlickr(this.props.photoId)
            .then(photoWrapper => {
                if (photoWrapper.isDateTakenKnown()) {
                    self.setState({
                        photoWrapper: photoWrapper
                    });
                }
            })
            .catch(error => console.log(error));
    }

    getPhotoDetailsFromFlickr(photoId) {
        const infoUrl = getPhotoInfoUrl(photoId);
        const infoUrlPromise = axios.get(infoUrl);

        return axios.all([infoUrlPromise])
            .then(axios.spread(function (infoResponse) {
                const photoWrapper = new FlickrPhotoWrapper(infoResponse.data);
                return photoWrapper;
            }));
    };

    render() {

        let componentToRender = null;
        let photoWrapper = this.state.photoWrapper;
        if (photoWrapper !== null) {
            componentToRender = <Photo id={photoWrapper.getId()}
                snapshotUrl={photoWrapper.getSnapshotUrl() } linkUrl={photoWrapper.getLinkUrl() }
                dateTakenFormatted={photoWrapper.getDateTakenAsPlainText() }
                title={photoWrapper.getTitle() } />;
        }

        return (componentToRender);
    }

};
