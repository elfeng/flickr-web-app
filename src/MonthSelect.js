import React, { Component } from 'react';
import { TWELVE_MONTHS } from './Month.js';

export default class MonthSelect extends Component {

    constructor(props) {
        super(props);
        this.selectMonth = this.selectMonth.bind(this);
    }

    selectMonth(monthNumber) {
        const month = TWELVE_MONTHS[monthNumber.target.value - 1];
        this.props.onChange(month);
    }

    render() {
        const monthOptions = TWELVE_MONTHS.map(month =>
            <option key={month.number} value={month.number}>{month.name}</option>
        );

        return (
            <select onChange={this.selectMonth} className="form-control" required
                id="month" name="month">
                <option value="">Month</option>
                {monthOptions}
            </select>
        );
    }

};
