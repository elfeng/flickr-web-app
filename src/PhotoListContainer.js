import axios from 'axios';
import React, { Component } from 'react';
import PhotoList from './PhotoList';
import PhotoContainer from './PhotoContainer';
import { getSearchUrl, NB_MAX_PHOTOS_PER_PAGE } from './FlickrURLs.js';

export default class PhotoListContainer extends Component {

    constructor(props) {
        super(props);
        this.photoContainers = [];
        this.state = {
            searchResults: null,
            photoContainers: []
        };
        this.searchFlickr = this.searchFlickr.bind(this);
        this.showMorePhotos = this.showMorePhotos.bind(this);
    }

    componentDidMount() {
        const self = this;
        this.searchFlickr(this.props.search, this.props.month, this.props.year)
            .then(searchResponse => {
                self.setState({
                    searchResults: searchResponse.data.photos.photo
                });
                this.showMorePhotos();
            })
            .catch(searchError => console.log(searchError));
    }

    searchFlickr(search, month, year) {
        const searchUrl = getSearchUrl(search, month, year);
        return axios.get(searchUrl);
    }

    /**
     * @return {number} the index of the last photo that will be displayed after we add more photos.
     */
    getNextLastPhotoIndex() {
        const lastPhotoIndex = this.state.photoContainers.length;
        const nbPhotosRemaining = this.state.searchResults.length - lastPhotoIndex;
        if (nbPhotosRemaining > NB_MAX_PHOTOS_PER_PAGE) {
            return lastPhotoIndex + NB_MAX_PHOTOS_PER_PAGE;
        } else if (nbPhotosRemaining > 0) {
            return lastPhotoIndex + nbPhotosRemaining;
        } else {
            return lastPhotoIndex;
        }
    }

    buildNewPhotoContainers() {
        const lastPhotoIndex = this.getNextLastPhotoIndex();
        for (let i = this.state.photoContainers.length; i < lastPhotoIndex; i++) {
            const photo = this.state.searchResults[i];
            this.photoContainers.push(<PhotoContainer key={photo.id} photoId={photo.id} />)
        }
    }

    showMorePhotos() {
        if (this.state.searchResults !== null) {
            this.buildNewPhotoContainers();
            this.setState({
                photoContainers: this.photoContainers
            });
        }
    }

    render() {

        let componentToRender = null;
        if (this.state.searchResults !== null) {
            componentToRender =
                <PhotoList
                    nbSearchResults={this.state.searchResults.length}
                    photoContainers={this.state.photoContainers}
                    showMorePhotos={this.showMorePhotos}/>;
        }

        return (componentToRender);
    }

};
