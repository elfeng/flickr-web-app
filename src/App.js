import React, {Component} from 'react';
import './App.css';
import MonthSelect from './MonthSelect';
import { YearSelect, getCurrentYear } from './YearSelect/YearSelect';
import PhotoListContainer from './PhotoListContainer';

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {};

    this.searchPhotos = this.searchPhotos.bind(this);
    this.onInputChange = this.onInputChange.bind(this);
    this.onMonthChange = this.onMonthChange.bind(this);
    this.onYearChange = this.onYearChange.bind(this);
  }

  onInputChange(search) {
    this.search = search.target.value;
  }

  onMonthChange(month) {
    this.month = month;
  }

  onYearChange(year) {
    this.year = year;
  }

  isDateSelectedAndNotInTheFuture() {
    return this.month && this.year &&
      (this.year < getCurrentYear() || this.month.isNotAfterCurrentMonth());
  }

  searchPhotos(event) {
    event.preventDefault();
    if (this.search && this.year && this.month) {
      this.setState({
        search: this.search,
        month: this.month,
        year: this.year
      });
    }
  }

  render() {

    const currentMillis = Date.now();
    let searchResults = null;
    if (this.state.search && this.isDateSelectedAndNotInTheFuture()) {
      searchResults = <PhotoListContainer key={currentMillis} search={this.state.search} month={this.state.month} year={this.state.year}/>;
    }

    return (
      <div className="App">

        <form onSubmit={this.searchPhotos}>
          <div className="container-fluid App-header">

            <div className="row">
              <div className="col-xs-12 col-sm-12 col-md-offset-2 col-md-10 col-lg-offset-2 col-lg-10 ">
                <h1>FlickrPicPicker:&nbsp;</h1>
                <h2>NASA Gallery Search by Month and Year</h2>
              </div>
            </div>

            <div className="row App-form" >
              <div className="col-xs-12 col-sm-12 col-md-offset-2 col-md-3 col-lg-offset-2 col-lg-3 ">
                <input type="text" onChange={this.onInputChange}
                  id="picalendr-search" name="picalendr-search"
                  className="form-control"
                  placeholder="Place, object or event"
                  autoFocus
                  required />
              </div>

              <div className="col-xs-6 col-sm-6 col-md-2 col-lg-2">
                <MonthSelect onChange={this.onMonthChange} />
              </div>

              <div className="col-xs-6 col-sm-6 col-md-2 col-lg-2">
                <YearSelect onChange={this.onYearChange}/>
              </div>

              <div className="col-xs-12 col-sm-12 col-md-1 col-lg-1">
                <input type="submit" className="btn btn-default" value="Search"/>
              </div>
            </div>
          </div>
        </form>

        <div className="container">

          <div className="row">
            <div className="col-xs-12 col-sm-12 col-md-offset-1 col-md-5 col-md-offset-1 col-lg-5 App-photos">
              {searchResults}
            </div>
            <div className="col-xs-12 col-sm-12 col-md-5 col-lg-5 App-intro">
              <h2>Welcome to FlickrPicPicker!</h2>
              <p>
                FlickrPicPicker is a search engine that allows you to search photographs by month, year and keyword, specifically from NASA's account.
                All the photos come from Flickr and are searchable as long as they exist under NASA's account at the time of search.
              </p>
              <h2>How does it work?</h2>
              <p>
                Pick a month and year you would like to retrieve NASA photos from. You can also enter a keyword, so that only photos whose title, description or tags contain the text will be returned.
              </p>
            </div>
          </div>
        </div>

      </div>
    );
  }
}

export default App;
