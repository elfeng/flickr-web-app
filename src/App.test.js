import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import App from './App';
import { getSearchUrl, getPhotoSnapshotUrl } from './FlickrURLs.js';
import { TWELVE_MONTHS } from './Month.js';
import renderer from 'react-test-renderer';
import MonthSelect from './MonthSelect';
import PhotoContainer from './PhotoContainer';
//import ReactTestUtils from 'react-addons-test-utils';
import ReactTestUtils from 'react-dom/test-utils';
import PhotoList from './PhotoList';
import PhotoListContainer from './PhotoListContainer';

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<PhotoListContainer month={TWELVE_MONTHS[0]} year={2017}/>, div);
});

describe("searchFlickr", () => {

  const march = TWELVE_MONTHS[2];
  let component = new PhotoListContainer();

  // it("should get photos for Tokyo in march 2014", () => {
  //   return component.searchFlickr("Tokyo", march, 2014).then((searchResponse) => {
  //     expect(searchResponse.data.photos.photo.length).not.toBe(0);
  //   });
  // })

  it("should not get photos for an unknown place", () => {
    return component.searchFlickr("X1e9T5C5P1f", march, 2014).then((searchResponse) => {
      expect(searchResponse.data.photos.photo.length).toBe(0);
    });
  })

});

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<PhotoContainer />, div);
});

describe("getPhotoDetailsFromFlickr", () => {

  let component = new PhotoContainer();

  it("should get the photo details given an existing id", () => {
    return component.getPhotoDetailsFromFlickr("25277702294").then((photoWrapper) => {
      expect(photoWrapper.isDateTakenKnown()).toBe(true);
    });
  })

  it("should not get the photo details given an unknown id", () => {
    return component.getPhotoDetailsFromFlickr("-1").then((photoWrapper) => {
      expect(photoWrapper.photo).toBe(undefined);
    });
  })

});

test('should match the snapshot', () => {
  const component = renderer.create(<MonthSelect />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('getSearchUrl should have dates', () => {
    const january = TWELVE_MONTHS[0];
    const january1st2016Seconds = 1451635200; // currentmillis.com
    const february1st2016Seconds = 1454313600; // currentmillis.com

    const searchUrl = getSearchUrl(" ", january, 2016);

    expect(searchUrl).not.toContain("undefined");
    expect(searchUrl).toContain(january1st2016Seconds);
    expect(searchUrl).toContain(february1st2016Seconds);
});

test('getPhotoSnapshotUrl', () => {
    let photoInfo = {
        farm: 1,
        server: 2,
        id: 3,
        secret: "secret4"
    };
    expect(getPhotoSnapshotUrl(photoInfo)).toBe("https://farm1.staticflickr.com/2/3_secret4.jpg");
});

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});
