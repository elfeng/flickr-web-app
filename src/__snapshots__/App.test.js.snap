// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<select
  className="form-control"
  id="month"
  name="month"
  onChange={[Function]}
  required={true}
>
  <option
    value=""
  >
    Month
  </option>
  <option
    value={1}
  >
    January
  </option>
  <option
    value={2}
  >
    February
  </option>
  <option
    value={3}
  >
    March
  </option>
  <option
    value={4}
  >
    April
  </option>
  <option
    value={5}
  >
    May
  </option>
  <option
    value={6}
  >
    June
  </option>
  <option
    value={7}
  >
    July
  </option>
  <option
    value={8}
  >
    August
  </option>
  <option
    value={9}
  >
    September
  </option>
  <option
    value={10}
  >
    October
  </option>
  <option
    value={11}
  >
    November
  </option>
  <option
    value={12}
  >
    December
  </option>
</select>
`;

exports[`should match the snapshot 2`] = `
<select
  className="form-control"
  id="month"
  name="month"
  onChange={[Function]}
  required={true}
>
  <option
    value=""
  >
    Month
  </option>
  <option
    value={1}
  >
    January
  </option>
  <option
    value={2}
  >
    February
  </option>
  <option
    value={3}
  >
    March
  </option>
  <option
    value={4}
  >
    April
  </option>
  <option
    value={5}
  >
    May
  </option>
  <option
    value={6}
  >
    June
  </option>
  <option
    value={7}
  >
    July
  </option>
  <option
    value={8}
  >
    August
  </option>
  <option
    value={9}
  >
    September
  </option>
  <option
    value={10}
  >
    October
  </option>
  <option
    value={11}
  >
    November
  </option>
  <option
    value={12}
  >
    December
  </option>
</select>
`;
